openmls/src/group/core_group/past_secrets.rs

```diff
@@ -14,47 +15,92 @@ struct EpochTree {
 /// Can store message secrets for up to `max_epochs`. The trees are added with [`self::add()`] and can be queried
 /// with [`Self::get_epoch()`].
 #[derive(Debug, Serialize, Deserialize)]
+#[cfg_attr(test, derive(PartialEq))]
 pub struct MessageSecretsStore {
+    // Maximum size of the `past_epoch_trees` list.
     max_epochs: usize,
-    epoch_trees: VecDeque<EpochTree>,
+    // Past message secrets.
+    past_epoch_trees: VecDeque<EpochTree>,
+    // The message secrets of the current epoch.
+    message_secrets: MessageSecrets,
 }
 
 impl MessageSecretsStore {
-    /// Create a new store that can hold up to `max_epochs` message secrets.
-    /// If `max_epochs` is 0, no secret trees will be stored.
-    pub fn new(max_epochs: usize) -> Self {
+    /// Create a new store that can hold up to `max_past_epochs` message secrets.
+    /// If `max_past_epochs` is 0, only the current epoch is being stored.
+    pub(crate) fn new_with_secret(max_epochs: usize, message_secrets: MessageSecrets) -> Self {
         Self {
             max_epochs,
-            epoch_trees: VecDeque::new(),
+            past_epoch_trees: VecDeque::new(),
+            message_secrets,
+        }
+    }
+
+    /// Resize the store.
+    pub(crate) fn resize(&mut self, max_past_epochs: usize) {
+        self.max_epochs = max_past_epochs;
+        while self.past_epoch_trees.len() > self.max_epochs {
```

Comment:
Trying to comment here again: You could avoid the while loop (which is always dangerous because it can loop indefinitely) by truncating from the front of the `VecDeque`. There is no dedicated function for that, but you can use `rotate_left(n)` and `truncate(n)`, where `n` is the number of elements you want to drop.